File: services/products/src/main/java/br/com/deltaglobalbank/products/infrastructure/persistence/adapters/ProductRepositoryAdapter.java
package br.com.deltaglobalbank.products.infrastructure.persistence.adapters;

import java.util.List;
import java.util.UUID;

import br.com.deltaglobalbank.products.domain.product.Product;
import br.com.deltaglobalbank.products.domain.product.ProductRepository;
import br.com.deltaglobalbank.products.domain.product.ProductType;
import br.com.deltaglobalbank.products.infrastructure.persistence.entities.ProductEntity;
import br.com.deltaglobalbank.products.infrastructure.persistence.mappers.ProductMapper;
import br.com.deltaglobalbank.products.infrastructure.persistence.repositories.JpaProductRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
public class ProductRepositoryAdapter implements ProductRepository {

    private final JpaProductRepository jpa;

    public ProductRepositoryAdapter(JpaProductRepository jpa) {
        this.jpa = jpa;
    }

    @Override
    @Transactional
    public Product save(Product product) {
        ProductEntity existing = jpa.findById(product.getId()).orElse(null);
        ProductEntity entity = existing != null
            ? ProductMapper.applyTo(product, existing)
            : ProductMapper.toEntity(product);
        jpa.save(entity);
        return findById(product.getId(), product.getTenantId());
    }

    @Override
    public Product findById(UUID id, UUID tenantId) {
        ProductEntity entity = jpa.findByIdAndTenantId(id, tenantId);
        return entity != null ? ProductMapper.toDomain(entity) : null;
    }

    @Override
    public boolean existsActiveByTypeAndAgreement(UUID tenantId, ProductType type, String agreementName) {
        return jpa.existsByTenantIdAndTypeAndAgreementNameIgnoreCaseAndActiveTrue(
            tenantId, type.toDatabaseValue(), agreementName);
    }

    @Override
    public Page<Product> findLendingPage(UUID tenantId, String agreementName, Boolean active, Pageable pageable) {
        return jpa.searchLending(tenantId, ProductType.LENDING.toDatabaseValue(), agreementName, active, pageable)
            .map(ProductMapper::toDomain);
    }

    @Override
    public boolean existsAnotherActiveWithAgreement(UUID tenantId, ProductType type, String agreementName, UUID excludeId) {
        return jpa.existsAnotherActiveWithAgreement(tenantId, type.toDatabaseValue(), agreementName, excludeId);
    }

    @Override
    public List<Product> findProducts(UUID tenantId, ProductType type, Boolean active, String agreementName) {
        return jpa.searchProducts(tenantId, type != null ? type.toDatabaseValue() : null, active, agreementName)
            .stream()
            .map(ProductMapper::toDomain)
            .toList();
    }
}

File: services/products/src/main/kotlin/br/com/deltaglobalbank/products/infrastructure/persistence/adapters/ProductRepositoryAdapter.kt
package br.com.deltaglobalbank.products.infrastructure.persistence.adapters

import br.com.deltaglobalbank.products.domain.product.Product
import br.com.deltaglobalbank.products.domain.product.ProductRepository
import br.com.deltaglobalbank.products.domain.product.ProductType
import br.com.deltaglobalbank.products.infrastructure.persistence.mappers.applyTo
import br.com.deltaglobalbank.products.infrastructure.persistence.mappers.toDomain
import br.com.deltaglobalbank.products.infrastructure.persistence.mappers.toEntity
import br.com.deltaglobalbank.products.infrastructure.persistence.repositories.JpaProductRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Component
class ProductRepositoryAdapter(private val jpa: JpaProductRepository) : ProductRepository {

    @Transactional
    override fun save(product: Product): Product {
        val existing = jpa.findById(product.id).orElse(null)
        val entity = if (existing != null) product.applyTo(existing) else product.toEntity()
        jpa.save(entity)
        return findById(product.id, product.tenantId)!!
    }

    override fun findById(id: UUID, tenantId: UUID): Product? =
        jpa.findByIdAndTenantId(id, tenantId)?.toDomain()

    override fun existsActiveByTypeAndAgreement(tenantId: UUID, type: ProductType, agreementName: String): Boolean =
        jpa.existsByTenantIdAndTypeAndAgreementNameIgnoreCaseAndActiveTrue(tenantId, type.toDatabaseValue(), agreementName)

    override fun findLendingPage(tenantId: UUID, agreementName: String?, active: Boolean?, pageable: Pageable): Page<Product> =
        jpa.searchLending(tenantId, ProductType.LENDING.toDatabaseValue(), agreementName, active, pageable)
            .map { it.toDomain() }

    override fun existsAnotherActiveWithAgreement(tenantId: UUID, type: ProductType, agreementName: String, excludeId: UUID) =
        jpa.existsAnotherActiveWithAgreement(tenantId, type.toDatabaseValue(), agreementName, excludeId)

    override fun findProducts(tenantId: UUID, type: ProductType?, active: Boolean?, agreementName: String?): List<Product> =
        jpa.searchProducts(tenantId, type?.toDatabaseValue(), active, agreementName).map { it.toDomain() }
}